experiments/Schema/src/Query.jl

```diff
@@ -36,301 +37,218 @@ This structure holds the relationship graph between fields in a query
-      n_edges[i] = (n_edges[i][1]+inc, n_edges[i][2]+inc)
-    end
-
-    # Make new query
-    n_query = Query(n_dom, n_cod, n_dom_names, n_cod_names, n_tables, n_fields, n_edges)
-    eval_variables!(n_query)
-    return n_query
-
+    n_tables = merge(f.tables, g.tables)
+    n_wd = otimes(f.wd, g.wd)
+    return Query(n_tables, n_wd)
   end
 
   meet(f::Query, g::Query) = begin
-    mcopy(dom(A))⋅(f⊗g)⋅mmerge(codom(f))
+    n_tables = merge(f.tables, g.tables)
+    n_wd = meet(f.wd, g.wd)
+    return Query(n_tables, n_wd)
   end
 
-  dagger(f::Query) = Query(f.codom, f.dom, f.codom_names, f.dom_names,
-                           f.tables, f.fields, f.edges)
+  dagger(f::Query) = Query(f.tables, dagger(f.wd))
 
   dunit(A::Types) = begin
-    create(A)⋅mcopy(A)
+    Query(dunit(A.ports))
   end
 
   top(A::Types, B::Types) = begin
-    delete(A)⋅create(B)
+    Query(top(A.ports,B.ports))
   end
 
   dcounit(A::Types) = begin
-    dagger(dunit(A))
+    Query(dcounit(A.ports))
   end
 
   id(A::Types) = begin
-    A_len = length(A.types)
-
-    domain   = collect(1:A_len)
-    codomain = collect((A_len+1):(2*A_len))
-
-    fields = fill((0,"*"),2*A_len)
-
-    edges = Array{Tuple{Int,Int},1}()
-    for i in 1:A_len
-      append!(edges, [(i,i+A_len)])
-    end
-
-    Query(A, A, domain, codomain, [],
-          fields,
-          edges)
+    Query(id(A.ports))
   end
 
   braid(A::Types, B::Types) = begin
-    A_len = length(A.types)
-    B_len = length(B.types)
-    tot_len = A_len + B_len
-
-    domain   = collect(1:(tot_len))
-    codomain = collect((tot_len+1):((tot_len)*2))
-
-    fields = fill((0,"*"),(tot_len)*2)
-
-    edges = Array{Tuple{Int,Int},1}()
-    for i in 1:length(domain)
-      if i > A_len
-        append!(edges, [(i,i-A_len + tot_len)])
-      else
-        append!(edges, [(i,i+B_len + tot_len)])
-      end
-    end
-
-    Query(otimes(A,B), otimes(B,A), domain, codomain, [],
-          fields, edges)
+    Query(braid(A.ports,B.ports))
   end
 
   mcopy(A::Types) = begin
-    A_len = length(A.types)
-    domain   = collect(1:A_len)
-    codomain = collect((A_len+1):(A_len*3))
-
-    edges = Array{Tuple{Int,Int},1}()
-    for i in 1:length(domain)
-      append!(edges, [(i,i+A_len),(i,i+A_len*2)])
-    end
-
-    fields = fill((0,"*"),A_len*3)
-
-    Query(A, otimes(A,A), domain, codomain, [],
-          fields, edges)
+    Query(implicit_mcopy(A.ports,2))
   end
 
   mmerge(A::Types) = begin
-    dagger(mcopy(A))
+    Query(implicit_mmerge(A.ports,2))
   end
 
   delete(A::Types) = begin
-    A_len = length(A.types)
-    domain   = collect(1:A_len)
-    codomain = Array{Int,1}()
-
-    fields = fill((0,"*"),A_len)
-    edges = Array{Tuple{Int,Int},1}()
-
-    Query(A, munit(Types), domain, codomain, [],
-          fields, edges)
+    Query(delete(A.ports))
   end
 
   create(A::Types) = begin
-    dagger(delete(A))
+    Query(create(A.ports))
   end
 end
 
 make_query(s::Schema, q::GATExpr)::Query = begin
```

Comment:
I think this should be a constructor of `Query` all you should have to do is change the name to Query.